core/src/main/scala/com/mesosphere/usi/core/matching/PortResource.scala

```diff
@@ -0,0 +1,223 @@
+    val offeredPorts = resourceToPortRange(resource)
+    if (offeredPorts.isEmpty || ports.isEmpty) {
+      return Seq.empty
+    }
+
+    // non-dynamic hostPorts
+    val staticPorts = ports.collect { case port if port != 0 => port }.toSet
+    val availablePortsWithoutStaticPorts: Iterator[Int] =
+      lazyRandomPortsFromRanges(offeredPorts, random).filter(port => !staticPorts(port))
+
+    val portMatchResults = ports.map {
+      case port if port == RandomPort && !availablePortsWithoutStaticPorts.hasNext =>
+        // need dynamic port but no more available
+        PortNotAvailable
+      case port if port == RandomPort && availablePortsWithoutStaticPorts.hasNext =>
+        // pick next available dynamic port
+        PortMatched(availablePortsWithoutStaticPorts.next())
+      case port if offeredPorts.exists(_.contains(port)) =>
+        // static port
+        PortMatched(port)
+      case _ =>
+        PortNotAvailable
+    }
+
+    if (portMatchResults.contains(PortNotAvailable)) {
+      Seq.empty
+    } else {
+      createPortsResources(resource, portMatchResults.collect { case PortMatched(port) => port }.toSeq, resourceType)
+    }
+  }
+
+  override def matchAndConsume(resources: Seq[Protos.Resource]): Option[ResourceMatchResult] = {
+    findResourceToMatch(Nil, resources.toList)
+  }
+
+  /**
+    * Converts given Resource to sequence of port ranges
+    * If given resource is not ports resource of type range, it will return empty sequence
+    * @param resource mesos resource definition
+    * @return list of port ranges in that resource
+    */
+  private def resourceToPortRange(resource: Protos.Resource): Seq[PortRange] = {
+    if (resource.getName != resourceType.name) {
+      Seq.empty
+    } else {
+      val rangeInResource = resource.getRanges.getRangeList.asScala
+      rangeInResource.map { range =>
+        PortRange(range.getBegin.toInt, range.getEnd.toInt)
+      }
+    }
+  }
+
+  /**
+    * We want to make it less likely that we are reusing the same dynamic port for tasks of different pods.
+    * This way we allow load balancers to reconfigure before reusing the same ports.
+    *
+    * Therefore we want to choose dynamic ports randomly from all the offered port ranges.
+    * We want to use consecutive ports to avoid excessive range fragmentation.
+    *
+    * The implementation idea:
+    *
+    * * Randomize the order of the offered ranges.
+    * * Now treat the ports contained in the ranges as one long sequence of ports.
+    * * We randomly choose an index where we want to start assigning dynamic ports in that sequence. When
+    *   we hit the last offered port with wrap around and start offering the ports at the beginning
+    *   of the sequence up to (excluding) the port index we started at.
+    * * The next range is determined on demand. That's why an iterator is returned.
+    */
+  private def lazyRandomPortsFromRanges(offeredPortRanges: Seq[PortRange], rand: Random): Iterator[Int] = {
```

Comment:
this is reused from marathon. It's up for discussion whether we need it. When I discussed this with Tim we just noted that the motivations described in this comment seems to be valid...